In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_excel("E:\MachineLearning\Data\Telco_customer_churn.xlsx")

In [3]:
df.shape

(7043, 33)

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

In [5]:
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [6]:
df.describe()

,Count,Zip Code,Latitude,Longitude,Tenure Months,Monthly Charges,Churn Value,Churn Score,CLTV
count,7043.0,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,1.0,93521.964646,36.282441,-119.798880,32.371149,64.761692,0.265370,58.699418,4400.295755
std,0.0,1865.794555,2.455723,2.157889,24.559481,30.090047,0.441561,21.525131,1183.057152
min,1.0,90001.000000,32.555828,-124.301372,0.000000,18.250000,0.000000,5.000000,2003.000000
25%,1.0,92102.000000,34.030915,-121.815412,9.000000,35.500000,0.000000,40.000000,3469.000000
50%,1.0,93552.000000,36.391777,-119.730885,29.000000,70.350000,0.000000,61.000000,4527.000000
75%,1.0,95351.000000,38.224869,-118.043237,55.000000,89.850000,1.000000,75.000000,5380.500000
max,1.0,96161.000000,41.962127,-114.192901,72.000000,118.750000,1.000000,100.000000,6500.000000


In [7]:
df.isnull().sum().sort_values(ascending=False)

Churn Reason         5174
CustomerID              0
Count                   0
State                   0
Country                 0
Zip Code                0
Lat Long                0
Latitude                0
City                    0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Longitude               0
Internet Service        0
Online Security         0
Device Protection       0
Online Backup           0
Streaming TV            0
Streaming Movies        0
Contract                0
Tech Support            0
Paperless Billing       0
Payment Method          0
Total Charges           0
Monthly Charges         0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df['CustomerID'].nunique()

7043

In [10]:
df['Total Charges'].value_counts().head(20)

Total Charges
20.2     11
         11
19.75     9
19.9      8
19.65     8
20.05     8
19.55     7
45.3      7
20.15     6
19.45     6
20.25     6
20.3      5
20.45     5
70.6      4
75.3      4
74.7      4
44.4      4
69.9      4
19.85     4
50.15     4
Name: count, dtype: int64

In [11]:
spaces_count = df['Total Charges'].astype(str).str.strip().eq('').sum()

In [12]:
pd.to_numeric(df['Total Charges'], errors='coerce').isna().sum()

np.int64(11)

In [13]:
categorical_cols = df.select_dtypes(include=['object']).columns

summary = pd.DataFrame({
    'Unique Values': df[categorical_cols].nunique(),
    'Missing Values': df[categorical_cols].isnull().sum()
})

summary

,Unique Values,Missing Values
CustomerID,7043,0
Country,1,0
State,1,0
City,1129,0
Lat Long,1652,0
Gender,2,0
Senior Citizen,2,0
Partner,2,0
Dependents,2,0
Phone Service,2,0


In [14]:
for col in categorical_cols:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- CustomerID ---
CustomerID
3668-QPYBK    1
9237-HQITU    1
9305-CDSKC    1
7892-POOKP    1
0280-XJGEX    1
             ..
2569-WGERO    1
6840-RESVB    1
2234-XADUH    1
4801-JZAZL    1
3186-AJIEK    1
Name: count, Length: 7043, dtype: int64

--- Country ---
Country
United States    7043
Name: count, dtype: int64

--- State ---
State
California    7043
Name: count, dtype: int64

--- City ---
City
Los Angeles       305
San Diego         150
San Jose          112
Sacramento        108
San Francisco     104
                 ... 
Milford             4
Calpine             4
Standish            4
Tulelake            4
Olympic Valley      4
Name: count, Length: 1129, dtype: int64

--- Lat Long ---
Lat Long
33.964131, -118.272783    5
34.059281, -118.30742     5
34.048013, -118.293953    5
34.062125, -118.315709    5
34.039224, -118.266293    5
                         ..
40.181279, -120.392967    4
39.672813, -120.456699    4
40.346634, -120.386422    4
41.813521, -121.492666    4
39.191

In [15]:
numerical_cols = df.select_dtypes(include=['number']).columns

df[numerical_cols].describe().T

,count,mean,std,min,25%,50%,75%,max
Count,7043.0,1.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000
Zip Code,7043.0,93521.964646,1865.794555,90001.000000,92102.000000,93552.000000,95351.000000,96161.000000
Latitude,7043.0,36.282441,2.455723,32.555828,34.030915,36.391777,38.224869,41.962127
Longitude,7043.0,-119.798880,2.157889,-124.301372,-121.815412,-119.730885,-118.043237,-114.192901
Tenure Months,7043.0,32.371149,24.559481,0.000000,9.000000,29.000000,55.000000,72.000000
Monthly Charges,7043.0,64.761692,30.090047,18.250000,35.500000,70.350000,89.850000,118.750000
Churn Value,7043.0,0.265370,0.441561,0.000000,0.000000,0.000000,1.000000,1.000000
Churn Score,7043.0,58.699418,21.525131,5.000000,40.000000,61.000000,75.000000,100.000000
CLTV,7043.0,4400.295755,1183.057152,2003.000000,3469.000000,4527.000000,5380.500000,6500.000000


In [16]:
pd.crosstab(df['Churn Label'], df['Churn Value'])

Churn Value,0,1
Churn Label,,
No,5174,0
Yes,0,1869


In [17]:
pd.crosstab(
    df['Churn Label'],
    df['Churn Reason'].isna()
)

Churn Reason,False,True
Churn Label,,
No,0,5174
Yes,1869,0


In [18]:
df.groupby('Churn Label')['Churn Reason'].apply(
    lambda x: x.notna().sum()
)

Churn Label
No        0
Yes    1869
Name: Churn Reason, dtype: int64

## Leakage Audit

| Column | Classification | Regression | Clustering | Reason |
|--------|----------------|------------|------------|--------|
| CustomerID | Drop | Drop | Drop | Identifier |
| Count | Drop | Drop | Drop | Constant feature |
| Country | Drop | Drop | Drop | Constant feature |
| State | Drop | Drop | Drop | Constant feature |
| Churn Label | Target | Drop | Drop | Classification target |
| Churn Value | Drop | Drop | Drop | Duplicate of Churn Label |
| Churn Score | Drop | Drop | Drop | Existing churn-related score |
| Churn Reason | Drop | Drop | Drop | Post-churn information |
| CLTV | Drop | Target | Consider | Regression target |

# Data Understanding Summary

## Dataset Overview

- Number of rows: 7043
- Number of columns: 33
- Duplicate rows: 0
- Unique customers: 7043

## Missing Values

- `Churn Reason` contains 5174 missing values.
- Missingness is associated with customers who did not churn.
- Other columns do not contain missing values at this stage.

## Data Quality Observations

- `Count` is constant with value 1 for all customers.
- `Country` is constant (`United States`).
- `State` is constant (`California`).
- `Total Charges` is stored as an object and contains blank string values.
- `CustomerID` is unique for every customer.

## Target Variables

### Classification Target
`Churn Label`

### Regression Target
`CLTV`

## Potential Leakage / Excluded Features

- `Churn Value`: duplicate representation of `Churn Label`.
- `Churn Score`: existing churn-related score; should not be used as an input feature for the baseline churn prediction model.
- `Churn Reason`: information available after/associated with churn and therefore excluded from churn prediction features.
- `CustomerID`: identifier, not a predictive feature.